File: src/Models/PromptSpecification.cs
using YamlDotNet.Serialization;

namespace PromptSpec.Models
{

    /// <summary>
    /// Represents a complete prompt specification as defined in YAML.
    /// </summary>
    public class PromptSpecification
    {
        /// <summary>
        /// A unique identifier for the prompt.
        /// </summary>
        [YamlMember(Alias = "name")]
        public string Name { get; set; } = string.Empty;

        /// <summary>
        /// The version of the prompt template.
        /// </summary>
        [YamlMember(Alias = "version")]
        public string Version { get; set; }

        /// <summary>
        /// A human-readable description of the prompt's purpose.
        /// </summary>
        [YamlMember(Alias = "description")]
        public string Description { get; set; }

        /// <summary>
        /// The system-level instructions for the model.
        /// </summary>
        [YamlMember(Alias = "systemMessage")]
        public string SystemMessage { get; set; }

        /// <summary>
        /// The main prompt content with placeholders for dynamic content.
        /// </summary>
        [YamlMember(Alias = "template")]
        public string Template { get; set; } = string.Empty;

        /// <summary>
        /// Common LLM parameters.
        /// </summary>
        [YamlMember(Alias = "parameters")]
        public LLMParameters Parameters { get; set; }

        /// <summary>
        /// Model-specific parameters as a flexible key-value map.
        /// </summary>
        [YamlMember(Alias = "modelConfig")]
        public Dictionary<string, object> ModelConfig { get; set; }

        /// <summary>
        /// Placeholder definitions for validation.
        /// </summary>
        [YamlMember(Alias = "placeholders")]
        public Dictionary<string, PlaceholderDefinition> Placeholders { get; set; }

        /// <summary>
        /// The desired output format (text, json, xml, etc.).
        /// </summary>
        [YamlMember(Alias = "outputFormat")]
        public string OutputFormat { get; set; }
    }
}

File: python/promptspec/models/prompt_specification.py
"""
Complete prompt specification as defined in YAML.
"""
from typing import Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict
from .llm_parameters import LLMParameters
from .placeholder_definition import PlaceholderDefinition


class PromptSpecification(BaseModel):
    """Complete prompt specification as defined in YAML."""
    
    model_config = ConfigDict(
        populate_by_name=True,
        alias_generator=lambda field_name: {
            'system_message': 'systemMessage',
            'llm_config': 'modelConfig',
            'output_format': 'outputFormat',
        }.get(field_name, field_name)
    )
    
    name: str = Field(..., description="Unique identifier for the prompt")
    version: Optional[str] = Field(None, description="Version of the prompt template")
    description: Optional[str] = Field(None, description="Human-readable description")
    system_message: Optional[str] = Field(None, description="System-level instructions")
    template: str = Field(..., description="Main prompt content with placeholders")
    parameters: Optional[LLMParameters] = Field(None, description="Common LLM parameters")
    llm_config: Optional[Dict[str, Any]] = Field(
        None, 
        description="Model-specific parameters as key-value map"
    )
    placeholders: Optional[Dict[str, PlaceholderDefinition]] = Field(
        None, 
        description="Placeholder definitions for validation"
    )
    output_format: Optional[str] = Field(
        None, 
        description="Desired output format (text, json, xml, etc.)"
    )